feat: support tachyon halo2

dongchangYoo opened the pull request:
In this PR, the backend of `kroma-prover` changes from `halo2` to `tachyon-halo2`.

dongchangYoo committed:
build: change rust toolchain to 1.70.0-nightly

dongchangYoo committed:
feat: add mode that operates with `TachyonProver`

dongchangYoo committed:
build: bump up halo2 to `halo2-with-tachyon`

dongchangYoo committed:
test(prover-server): increase `CLIENT_TIMEOUT_SEC` of mock_client

The client timeout has been extended to accommodate the proof generation time.

chokobole commented on zkevm/Cargo.toml:
```diff
@@ -6,23 +6,23 @@ edition = "2021"
 is-even = "1.0.0"
 ethers-core = "0.17.0"
-sha2 ="0.10.2"
+sha2 = "0.10.2"
```
Please remove this from the commit!

chokobole commented:
`refac(bin): refactoring bin/setup.rs` this should be categorized into `feat` I guess.

chokobole commented:
Could you explain why the prover needs to create a param..? How about downloading the params that we want from the web?

chokobole commented on bin/src/prove.rs:
```diff
@@ -1,86 +1,81 @@
-                traces.insert(path.file_stem().unwrap().to_os_string(), block_trace);
-            }
+    fn get_proof_type(&self) -> CircuitType {
+        if &self.circuit == "evm" || &self.circuit == "Evm" || &self.circuit == "EVM" {
```
Please compare `self.circuit` in lower case

dongchangYoo commented:
> Could you explain why the prover needs to create a param..? How about downloading the params that we want from the web?

chokobole commented on bin/src/prove.rs:
```diff
@@ -28,6 +28,9 @@ struct Args {
     circuit: String,
+    /// Specify whether to create `Verifier.sol`. (default: true)
+    #[clap(short, long = "solidity")]
+    sol: Option<bool>,
```
Why is this `Option<bool>` instead of `bool`?

chokobole commented on bin/src/prove.rs:
```diff
@@ -165,17 +181,20 @@ fn main() {
+                    );
+                    log::info!("verifier solidity to {}", out_dir.to_str().unwrap());
+                }
                 log::info!("output files to {}", out_dir.to_str().unwrap());
```
Please remove this

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -292,6 +301,7 @@ impl Prover {
         fn from_0_to_n<const N: usize>() -> [usize; N] {
             core::array::from_fn(|i| i)
         }
+
```
Please remove this

dongchangYoo commented:
> Could you explain why the prover needs to create a param..? How about downloading the params that we want from the web?

The functionality that was in setup.rs is maintained, but as you said, it does not generate params at the production level. However, the feature to download params is implemented in download.sh.

chokobole commented on bin/src/prove.rs:
```diff
@@ -44,6 +47,13 @@ impl Args {
         return true;
     }
 
+    fn does_create_verifier_sol(&self) -> bool {
+        match &self.sol {
+            Some(sol) => sol.clone(),
+            _ => false,
+        }
+    }
```
If the type of `self.sol` is changed to `bool`, then this function seems not needed.

chokobole commented on ffi/src/prove.rs:
```diff
@@ -25,7 +25,7 @@ pub unsafe extern "C" fn create_agg_proof(trace_char: *const c_char) -> *const c
         .get_mut()
         .unwrap()
-        .create_agg_circuit_proof(&trace)
+        .create_agg_circuit_proof(&trace, true)
```
Why this is true and one in `prover-server/src/prove.rs` is false?

chokobole commented on prover-server/src/prove.rs:
```diff
@@ -124,6 +122,7 @@ pub fn write_agg_proof(dir: &Path, proof: &AggCircuitProof) {
     proof.write_to_dir(&mut proof_path);
 }
 
+#[allow(dead_code)]
```
Do we need this?

dongchangYoo replied:
I think it's redundant code. I decided to remove it.

chokobole commented on bin/src/setup.rs:
```diff
@@ -1,29 +1,50 @@
-    #[clap(short, long = "seed")]
-    seed_path: Option<String>,
+    /// Specify directory which params have stored in. (default: ./kzg_params)
+    #[clap(short, long = "params_dir")]
```
Please set `kzg_params` to the `params_dir` by default with a clap macro

chokobole commented on bin/src/setup.rs:
```diff
@@ -1,29 +1,50 @@
+                dir
+            }
+        }
+    }
```
Please remove this, use a clap default value.

chokobole commented:
Please change `feat(bin)` to `feat` in 17bb7dc84e5fbd827d441d2aba904ba5519cb663

chokobole commented on bin/src/prove.rs:
```diff
@@ -1,86 +1,81 @@
+        [
+            0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06,
+            0xbc, 0xe5,
+        ]
```
Please use the constant value from `halo2`

chokobole commented on bin/src/prove.rs:
```diff
@@ -37,6 +37,13 @@ enum CircuitType {
+        return false;
+        #[cfg(feature = "tachyon")]
+        return true;
+    }
```
This is not a proper location to implement `is_tachyon()`.

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -102,7 +118,10 @@ impl AggCircuitProof {
 pub struct Prover {
     pub params: ParamsKZG<Bn256>,
     pub agg_params: ParamsKZG<Bn256>,
+    #[cfg(not(feature = "tachyon"))]
     pub rng: XorShiftRng,
+    #[cfg(feature = "tachyon")]
+    pub rng: XORShiftRng,
```
Please use generic!

dongchangYoo replied:
If this is changed to Generic, you cannot put rng into `create_tachyon_proof`.

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -111,11 +130,24 @@ pub struct Prover {
+        Self {
+            params,
+            agg_params,
+            rng: rng,
```
Please use generic

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -153,6 +186,15 @@ impl Prover {
+        rng: XORShiftRng,
+    ) -> Self {
+        Self::new(params, agg_params, rng)
+    }
```
Please use generic

chokobole commented on bin/src/prove.rs:
```diff
@@ -28,6 +28,9 @@ struct Args {
     circuit: String,
+    /// Specify whether to create `Verifier.sol`. (default: true)
+    #[clap(short, long = "solidity", default_value_t = true)]
+    does_gen_sol: bool,
```
```suggestion
    #[clap(short, long = "gen_sol", default_value_t = true)]
    gen_sol: bool,
```

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -143,7 +147,11 @@ impl Prover {
+        params: ParamsKZG<Bn256>,
+        agg_params: Option<ParamsKZG<Bn256>>,
+        rng: XORShiftRng,
+    ) -> Self {
```
Please use generic

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -224,7 +226,14 @@ impl Prover {
+        Self::from_params_and_seed(params, Some(agg_params), seed)
+    }
+
+    pub fn get_agg_params(&self) -> &ParamsKZG<Bn256> {
+        match &self.agg_params {
+            Some(params) => &params,
+            _ => panic!("agg params must be available"),
+        }
```
Please remove this

dongchangYoo replied:
By using this function, since agg_params are released, it becomes possible to quickly discover problems that may occur if used.

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -307,7 +317,7 @@ impl Prover {
 
             verify_vk: self.agg_pk.as_ref().expect("pk should be inited").get_vk(),
-            verify_params: &self.agg_params,
+            verify_params: self.get_agg_params(),
```
```suggestion
            verify_params: self.agg_params.as_ref().unwrap(),
```

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -375,11 +388,12 @@ impl Prover {
             log::info!("generate agg pk: begin");
-            let verify_circuit_vk =
-                keygen_vk(&self.agg_params, &verify_circuit).expect("keygen_vk should not fail");
+            let verify_circuit_vk = keygen_vk(self.get_agg_params(), &verify_circuit)
```
```suggestion
            let verify_circuit_vk = keygen_vk(self.agg_params.as_ref().unwrap(), &verify_circuit)
```

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -375,11 +388,12 @@ impl Prover {
-            let verify_circuit_pk = keygen_pk(&self.agg_params, verify_circuit_vk, &verify_circuit)
-                .expect("keygen_pk should not fail");
+            let verify_circuit_pk =
+                keygen_pk(self.get_agg_params(), verify_circuit_vk, &verify_circuit)
```
```suggestion
                keygen_pk(self.agg_params.as_ref().unwrap(), verify_circuit_vk, &verify_circuit)
```

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -426,10 +440,15 @@ impl Prover {
             let mut prover = {
                 let mut params_bytes = vec![];
-                self.agg_params.write(&mut params_bytes).unwrap();
+                let agg_params_ref = self.get_agg_params();
```
```suggestion
                let agg_params_ref = self.agg_params.as_ref().unwrap();
```

dongchangYoo committed:
feat(bin): remove `seed_path` option

The option related to `seed` was removed since it is used only for test.
We will consider it fixed value in development stage.

dongchangYoo committed:
feat(bin): update the `params_path`

rename `params_path` to `params_dir`.
The name indicates that 'params' must necessarily be a directory.

`params_dir` will be set to `./kzg_params` when it omitted.

dongchangYoo committed:
feat: add `n` option to setup.rs

User can generate `params` about `n` specified from command line.
If the `n` option is omitted, `setup` generates 2 params about `DEGREE`
and `AGG_DEGREE`.

dongchangYoo committed:
feat(bin): change the `trace_path` to a required option

This is because `prove.rs` does not operate without a trace.

dongchangYoo committed:
feat(bin): change type of `circuit` option to enum

dongchangYoo committed:
refac(bin): refactor `prover.rs`

chokobole commented on bin/src/prove.rs:
```diff
@@ -49,10 +49,9 @@ fn main() {
-        .expect("failed to load kzg params");
+    let params = load_kzg_params(&args.params_dir, *DEGREE).expect("failed to load kzg params");
+    let agg_params =
+        load_kzg_params(&args.params_dir, *AGG_DEGREE).expect("failed to load kzg params");
```
```suggestion
        load_kzg_params(&args.params_dir, *AGG_DEGREE).expect("failed to load kzg agg params");
```

chokobole commented on bin/src/prove.rs:
```diff
@@ -1,3 +1,4 @@
+use clap::ArgEnum;
 use clap::Parser;
```
```suggestion
use clap::{ArgEnum, Parser};
```

chokobole commented on bin/src/prove.rs:
```diff
@@ -25,18 +34,9 @@ struct Args {
-    /// Boolean means if output agg proof.
-    #[clap(long = "agg")]
-    agg_proof: Option<bool>,
+    /// Specify circuit type in [evm, state, agg]
```
```suggestion
    /// Specify circuit type in [evm, state, agg].
```

chokobole commented on zkevm/src/prover.rs:
```diff
@@ -111,11 +130,32 @@ pub struct Prover {
+        Self {
+            params,
+            agg_params,
+            rng: rng,
```
```suggestion
            rng,
```

chokobole commented on zkevm/Cargo.toml:
```diff
@@ -47,6 +47,7 @@ procfs = "0.13.0"
 default = []
 #default = ["prove_verify"]
 prove_verify = []
+tachyon = []
```
Please remove this line since I think this commit should only contain bump up part.

chokobole commented on types/Cargo.toml:
```diff
@@ -4,7 +4,7 @@ version = "0.3.0"
 
 [dependencies]
-eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "6ef08f5", feature = ["kroma"]}
+eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf" }
```
```suggestion
eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf", feature = ["kroma"] }
```

chokobole commented on zkevm/Cargo.toml:
```diff
@@ -6,17 +6,17 @@ edition = "2021"
 
-bus-mapping = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "6ef08f5", feature = ["kroma"] }
-eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "6ef08f5", feature = ["kroma"] }
+bus-mapping = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf" }
+eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf" }
```
```suggestion
bus-mapping = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf" feature = ["kroma"] }
eth-types = { git = "https://github.com/kroma-network/zkevm-circuits.git", rev = "4f431bf" feature = ["kroma"] }
```

chokobole commented on bin/Cargo.toml:
```diff
@@ -20,9 +20,10 @@ serde = "1.0"
 serde_json = "1.0.66"
 tokio = { version = "1", features = ["full"] }
-zkevm = { path = "../zkevm" }
+zkevm = { path = "../zkevm", features = ["tachyon"] }
```
Why does it need `features = ["tachyon"]`?

```suggestion
zkevm = { path = "../zkevm" }
```

chokobole commented on bin/src/prove.rs:
```diff
@@ -19,9 +19,9 @@ use zkevm::{
-    #[clap(short, long = "params")]
-    params_path: Option<String>,
+    /// Specify directory which params have stored in. (default: ./kzg_params)
+    #[clap(default_value = "./kzg_params", short, long = "params_dir")]
```
```suggestion
    #[clap(default_value = "./kzg_params", short, long)]
```

chokobole commented on bin/src/setup.rs:
```diff
@@ -4,9 +4,9 @@ use zkevm::{circuit::DEGREE, utils::load_or_create_params};
-    #[clap(short, long = "params")]
-    params_path: Option<String>,
+    /// Specify directory which params have stored in. (default: ./kzg_params)
+    #[clap(default_value = "./kzg_params", short, long = "params_dir")]
```
```suggestion
    #[clap(default_value = "./kzg_params", short, long)]
```

chokobole commented on bin/src/setup.rs:
```diff
@@ -1,18 +1,32 @@
 struct Args {
     /// Specify directory which params have stored in. (default: ./kzg_params)
-    #[clap(default_value = "./kzg_params", short, long = "params_dir")]
+    #[clap(default_value = "./kzg_params", long = "params_dir")]
```
Please don't remove `short` in this commit!

chokobole commented on bin/src/prove.rs:
```diff
@@ -22,9 +22,9 @@ struct Args {
-    #[clap(short, long = "trace")]
-    trace_path: Option<String>,
+    /// Specify path to block trace. (json file or directory)
+    #[clap(short, long = "trace_path")]
```
```suggestion
    #[clap(short, long)]
```

chokobole commented on bin/src/prove.rs:
```diff
@@ -25,18 +33,9 @@ struct Args {
-    #[clap(long = "agg")]
-    agg_proof: Option<bool>,
+    /// Specify circuit type in [evm, state, agg].
+    #[clap(short, long = "circuit", arg_enum)]
```
```suggestion
    #[clap(short, long, arg_enum)]
```

chokobole commented on prover-server/Cargo.toml:
```diff
@@ -44,4 +44,6 @@ name = "mock-client"
 path = "src/mock_client.rs"
 
 [features]
+default = ["tachyon"]
+tachyon = []
```
Please remove these since they are not used.

chokobole commented on bin/Cargo.toml:
```diff
@@ -25,6 +25,10 @@ types = { path = "../types", features = ["test"] }
+[features]
+default = ["tachyon"]
+tachyon = []
+
```
Please remove these since they are not used.

chokobole commented on bin/Cargo.toml:
```diff
@@ -25,6 +25,10 @@ types = { path = "../types", features = ["test"] }
 utils = { path = "../utils" }
```
```suggestion
utils = { path = "../utils", features = ["tachyon"] }
```

chokobole commented on prover-server/Cargo.toml:
```diff
@@ -23,7 +23,7 @@ serde = "1.0.164"
 types = { path = "../types", features = ["test"] }
-zkevm = { path = "../zkevm" }
+zkevm = { path = "../zkevm", features = ["tachyon"] }
 utils = { path = "../utils" }
```
```suggestion
utils = { path = "../utils", features = ["tachyon"] }
```

dongchangYoo committed:
style: rearrange importing orders of crates

chokobole reviewed (approved):
LGTM
